File: decryptors/decrypt.py
import base64
import hashlib
import json
import os
import time
from typing import Any, ByteString, Dict

import requests
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

APIKEY = os.environ["APIKEY"]
SECRET = os.environ["SECRET"]

TRANSIT_TIME_BUCKET = os.environ.get("TRANSIT_TIME_BUCKET", 60)
TRANSIT_KEY_LENGTH = os.environ.get("TRANSIT_KEY_LENGTH", 60)
HOST = os.environ.get("HOST", "0.0.0.0")
PORT = os.environ.get("PORT", 8080)


def transit_decrypt(ciphertext: str | ByteString) -> Dict[str, Any]:
    """Decrypt transit encrypted payload."""
    epoch = int(time.time()) // TRANSIT_TIME_BUCKET
    hash_object = hashlib.sha256(f"{epoch}.{APIKEY}.{SECRET}".encode())
    aes_key = hash_object.digest()[:TRANSIT_KEY_LENGTH]
    if isinstance(ciphertext, str):
        ciphertext = base64.b64decode(ciphertext)
    decrypted = AESGCM(aes_key).decrypt(ciphertext[:12], ciphertext[12:], b"")
    return json.loads(decrypted)


def get_cipher() -> str:
    """Get ciphertext from the server."""
    headers = {
        "accept": "application/json",
        "Authorization": f"Bearer {APIKEY}",
    }
    params = {
        "table_name": "default",
    }
    response = requests.get(
        f"http://{HOST}:{PORT}/get-table",  # noqa: HttpUrlsUsage
        params=params,
        headers=headers,
    )
    assert response.ok, response.text
    return response.json()["detail"]


print(transit_decrypt(ciphertext=get_cipher()))

File: decryptors/decrypt.js
const crypto = require('crypto');
const axios = require('axios');

const APIKEY = process.env.APIKEY;
const SECRET = process.env.SECRET;

if (!(APIKEY || SECRET)) {
    console.error('APIKEY and SECRET environment variables must be set');
    process.exit(1);
}

const getEnvAsInt = (key, defaultValue) => {
    const value = process.env[key];
    return value !== undefined ? parseInt(value, 10) : defaultValue;
};

const TRANSIT_TIME_BUCKET = getEnvAsInt("TRANSIT_TIME_BUCKET", 60);
const TRANSIT_KEY_LENGTH = getEnvAsInt("TRANSIT_KEY_LENGTH", 60);
const HOST = process.env.HOST || "0.0.0.0";
const PORT = getEnvAsInt("PORT", 8080);


function generateAuthHeader(token) {
    const timestamp = String(Math.floor(Date.now() / 1000));
    const signature = crypto.createHmac('sha512', token).update(timestamp).digest('hex');
    return `Signature=${signature},timestamp=${timestamp}`;
}

async function transitDecrypt(ciphertext) {
    const epoch = Math.floor(Date.now() / (1000 * TRANSIT_TIME_BUCKET));
    const hash = crypto.createHash('sha256');
    hash.update(`${epoch}.${APIKEY}.${SECRET}`);
    const aesKey = hash.digest().slice(0, TRANSIT_KEY_LENGTH);

    const bufferCiphertext = Buffer.from(ciphertext, 'base64');
    if (bufferCiphertext.length < 12 + 16) {
        throw new Error('Ciphertext is too short');
    }

    const iv = bufferCiphertext.slice(0, 12); // First 12 bytes
    const authTag = bufferCiphertext.slice(bufferCiphertext.length - 16); // Last 16 bytes
    const encryptedData = bufferCiphertext.slice(12, bufferCiphertext.length - 16); // Data in between

    const decipher = crypto.createDecipheriv('aes-256-gcm', aesKey, iv);
    decipher.setAuthTag(authTag); // Set the authentication tag

    let decrypted;
    try {
        decrypted = Buffer.concat([decipher.update(encryptedData), decipher.final()]);
    } catch (err) {
        throw new Error('Decryption failed: ' + err.message);
    }

    return JSON.parse(decrypted.toString());
}

async function getCipher() {
    const headers = {
        'accept': 'application/json',
        'Authorization': `Bearer ${generateAuthHeader(APIKEY)}`,
    };
    const params = {
        table_name: 'default',
    };
    const response = await axios.get(`http://${HOST}:${PORT}/get-table`, {params, headers});  // noqa: HttpUrlsUsage
    if (response.status !== 200) {
        throw new Error(response.data);
    }
    return response.data.detail;
}

async function main() {
    try {
        const ciphertext = await getCipher();
        const result = await transitDecrypt(ciphertext);
        console.log(result);
    } catch (error) {
        console.error(error);
    }
}

main();
